tests/scipy_optimize_test.py

```diff
@@ -0,0 +1,259 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from absl.testing import absltest
+
+import jax
+from jax import numpy as jnp
+from jax import test_util as jtu
+import jax.scipy.sparse.linalg
+from jax.config import config
+from jax.scipy.optimize.bfgs_minimize import BFGSResults, bfgs_minimize
+from jax.scipy.optimize.line_search import LineSearchResults, line_search
+
+config.parse_flags_with_absl()
+
+from jax.config import config
+
+config.parse_flags_with_absl()
+
+
+def line_search_nojit(value_and_gradient, position, direction, f_0=None, g_0=None, max_iterations=50, c1=1e-4, c2=0.9):
+    """
+        Performs an inexact line-search. It is a modified backtracking line search. Instead of reducing step size by a
+        number < 1 if Wolfe conditions are not met, we check the sign of  u = del_t restricted_func(t).
+        If u > 0 then we do normal backtrack, otherwise we search forward. Normal backtracking can fail to satisfy strong
+        Wolfe conditions. This extra step costs one extra gradient evaluation. For explaination see figures 3.1--3.4 in
+        https://pages.mtu.edu/~struther/Courses/OLD/Sp2013/5630/Jorge_Nocedal_Numerical_optimization_267490.pdf
```

Comment:
Please give a full reference for Nocedal's book, rather than the URL for a possibly unauthorized copy on a random academic course website.

Reply:
Done.